scripts/web-vnc

```diff
@@ -0,0 +1,239 @@
+
+        x11vnc_command+=" -sid '${WINDOW_ID}'"
+    fi
+    x11vnc_command+=" &"
+    echo "Starting x11vnc using port ${VNC_PORT}..."
+    eval "${x11vnc_command}"
+
+    echo "x11vnc PID is $!"
+    PIDS+=("$!")
+}
+
+
+start_novnc() {
+    echo "Starting novnc server using http port ${HTTP_PORT}..."
+    novnc_command="/usr/share/novnc/utils/launch.sh  --listen ${HTTP_PORT} --vnc localhost:${VNC_PORT}"
+    if [ -f "${SSL_CERT}" ]; then
+        echo "Using certificate in ${SSL_CERT}"
+        novnc_command+=" --cert ${SSL_CERT}"
+    fi
+    novnc_command+=" &"
+    eval "${novnc_command}"
+
+    echo "novnc PID is $!"
+    PIDS+=("$!")
+
+    sleep 2
+}
+
+
+kill_pids() {
+    for PID in "${PIDS[@]}"; do
+        echo "Killing PID ${PID}..."
+        if [ -n "${PID}" ]; then
+            kill "${PID}" || true
+        fi
+    done
+}
+
+
+create_pid_file() {
+    echo "Storing PIDs into ${PIDS_FILE}"
+    printf "%s\n" "${PIDS[@]}" > "${PIDS_FILE}"
+}
+
+
+start_window_manager() {
+    eval "DISPLAY=:${DISPLAY_ID} startlxde-pt &"
+    PIDS+=("$!")
+}
+
+
+run_on_start_command() {
+    if [ -n "${RUN_COMMAND}" ]; then
+        start_window_manager
+
+        eval "DISPLAY=:${DISPLAY_ID} ${RUN_COMMAND} &"
+        PIDS+=("$!")
+    fi
+}
+
+
+do_start() {
+    do_stop
+
+    # Create virtual display if necessary
+    start_xvfb || (kill_pids && exit 1)
+
+    # Run provided command
+    run_on_start_command
+
+    # Start VNC & novnc
+    start_x11vnc || (kill_pids && exit 1)
+    start_novnc || (kill_pids && exit 1)
+
+    create_pid_file
+}
+
+
+do_stop() {
+    if [ -f "${PIDS_FILE}" ]; then
+        # Kill PIDs of processes associated with DISPLAY_ID
+        echo "Found PIDs file ${PIDS_FILE}"
+        while read PID ; do
+            PIDS+=("$PID")
+        done < "${PIDS_FILE}"
+
+        kill_pids
+
+        echo "Removing ${PIDS_FILE}"
+        rm -r "${PIDS_FILE}"
+    else
+        echo "No PID file for display ${DISPLAY_ID}, skipping..."
+    fi
+}
+
+
+get_status() {
+    #ls /tmp/.pt-os-web-portal-vnc.*.pid | grep -e [0-9][0-9][0-9] -o
+    return
+}
+
+
+get_url() {
+    echo "http://pi-top.local:${HTTP_PORT}/vnc.html?autoconnect=true"
+}
+
+
+usage() {
+  echo -e "Usage:\n\tpt-os-web-portal-vnc <COMMAND> --display_id <DISPLAY_ID> --ssl_certificate <SSL_CERTIFICATE> --window_title <WINDOW_TITLE> --run <RUN_COMMAND>"
+  echo "where:"
+  echo -e "\tCOMMAND: start, stop, url."
+  echo -e "\tDISPLAY_ID: integer, id for the display to use/create."
+  echo -e "\tSSL_CERTIFICATE: path to combined SSL certificate & key file. Optional."
+  echo -e "\tWINDOW_TITLE: Title of a window in a display to share over VNC. Optional."
+  echo -e "\tRUN_COMMAND: Command to run before starting VNC server."
+}
+
+
+PIDS=""
+DISPLAY_ID="0"
+SSL_CERT=""
+WINDOW_TITLE=""
+RUN_COMMAND=""
+
+
+# Parse options
+VALID_ARGS=$(getopt -o w:i:s:r:h --long window_title:,display_id:,ssl_certificate:,run:,help -- "$@")
+if [[ $? -ne 0 ]]; then
+    exit 1;
+fi
```

Comment:
this is all very fancy and nice (further-link-vnc will be sad 😢), but there are a couple of common cases where we should be printing usage but don't:

```
pi@pi-top:~ $ /usr/bin/web-vnc
/usr/bin/web-vnc: line 235: 1: unbound variable
pi@pi-top:~ $ /usr/bin/web-vnc -d
getopt: invalid option -- 'd'
```